### ISSUE TITLE
Clarify Decision to Use DOMPoints instead of TypedArrays for XRViewerPose Properites?

### ISSUE THREAD

**Issue by KTRosenberg** [immersive-web/webxr#946]
In short, I am curious where the decision to use `DOMPoint`s to represent `XRViewerPose` position and orientation (and others) has been documented. I've looked through the issues but found no discussion. I would argue in favor of somehow supporting both JSON-like and array representations of data to make working with different APIs easier.

For a little context, I am porting a WebGL / WebVR backend to use WebXR. In most engine code I've written that interacts with graphics APIs, I find it most convenient to represent data in ways that let me easily upload to the GPU, i.e. by using Float32Arrays in the case of JavaScript/WebGL code so no conversion is necessary.

I can see why `DOMPoint // {x : #, y: #, z#, etc.}`s are convenient. For example, there is no question about where each component of a quaternion is stored (xyzw, wxyz, etc.). The popular THREE.js uses this format. Readability is arguably the advantage here.

However, using a typed array allows the programmer to funnel data directly to the GPU without additional conversion costs (assume microbenchmarking is okay here). DOMPoints or similar object representations aren't raw data, which means conversion to and from raw data is necessary when interacting with the GPU or APIS that use arrays. It seems that the assumption is that position/orientation will be used strictly in calculations, and will not necessarily be uploaded to the GPU, but other popular libraries like gl-matrix https://github.com/toji/gl-matrix use the Array format for its math. (Based on the various issues and pull requests I've read here, I believe the glMatrix repository owner is in the WebXR group, so I am sure there was some discussion about the topic at hand.)

For my project, I have already found it necessary to create wrappers to support both representations of positions and quaternions, like so:

```js

// this is a wrapper around XRViewerPose that provides additional
// information, namely, pre-computed buffer versions of
// position and orientation. Update with an explicit function call, per frame
export class XRViewerPoseEXT {
    constructor() {
        this.viewerPose         = null;
        this.positionAsArray    = new Float32Array(3);
        this.orientationAsArray = new Float32Array(4);
    }
    
    update(viewerPose) {
        this.viewerPose = viewerPose;

        const xform = this.viewerPose.transform;
        positionObjectToArray(xform.position, this.positionAsArray);
        orientationObjectToArray(xform.orientation, this.orientationAsArray);
    }

    isValid() {
        return (this.viewerPose != null);
    }
}
```
Although creating wrappers like the above is "fine" minus the additional time spent transferring data to buffers, I figure there might be a better way of supporting both the "readability-oriented" libraries and the "data-oriented" libraries already being used in the wild.

Assuming the WebXR spec and implementations can easily adapt to this, is it possible to support both representations without any additional calculation costs or branches at the lower level? For example, could WebXR be configured at initialization time to provide the raw data versions of position and orientation? I suspect this might be possible because the position and orientation (and other DOMPoint data) needs to be calculated and stored as x,y,z,etc. properties somehow. Without knowing much about the underlying implementations, could it be that there are already array buffers that could simply be exposed as properties?

I would be interested in learning more about this decision and seeing if perhaps WebXR could offer more of a choice in the matter in future versions, especially in the advent of more performance-minded APIs like WebGPU coming eventually. Could this be up for discussion for future revisions and additions to the WebXR API?

**Comment by klausw** [comment]
> Without knowing much about the underlying implementations, could it be that there are already array buffers that could simply be exposed as properties?

In the Chrome implementation, the poses are transferred via IPC and have varying internal representations in different parts of the code, so there are conversions going on already. There aren't currently specialized array accessors for position/orientation, but poses have a _XRRigidTransform_ which includes a [matrix](https://developer.mozilla.org/en-US/docs/Web/API/XRRigidTransform/matrix) property, and that's a Float32Array. Depending on your graphics library, using the matrix representation may well be more convenient, especially if your goal is to put the data into a GL uniform for use by the GPU. Alternatively, you could use _DOMMatrixReadOnly_'s _{to,from}Float32Array_ methods to work with them using DOM APIs.

While I don't know the full background of the decision, making sure that the representation is unambiguous and well-documented was an important criterion, along with the goal of not inventing new abstractions when existing ones such as _DOMPointReadOnly_ exist.

Currently, the WebXR core APIs only include a handful of poses that are transferred this way, so efficiency shouldn't be a major concern. Going forward, for new APIs such as meshes for real world understanding, I think people are looking into supporting more efficient packed array representations or (for large data) potentially keeping them on the GPU, but that's an area of ongoing discussion.

**Comment by KTRosenberg** [comment]
I see. It makes sense that the underlying data representation would be non-standard. Using the `XRRigidTransform` matrix might be more efficient in some cases as you say, but I'd still need to do some extra calculations to extract individual position and orientation data. Also, speaking from a data-packing perspective, you could send position/orientation data as 7 numbers instead of 16 if the position and orientation attributes were available.

It's not a huge dilemma performance-wise, of course, but it definitely would be nice to have the option of using buffers directly if I wanted to avoid creating wrappers such as the one above.

In any case, thank you for clarifying.

**Comment by toji** [comment]
For further reference, the PR where this was originally discussed and added is https://github.com/immersive-web/webxr/pull/357, and the comments there provide a bit of background on where the decision came from. Unfortunately a portion of the discussion occurred at a face to face meeting which I'm having trouble finding the minutes for. My recollection of the primary decision drivers was simply that we couldn't predict every way that developers would need to use the data so we wanted it to be easy to understand how you interacted with the data with as little context as possible. Additionally, the desire to avoid creating specialized new vector types was strong. `DOMPoint` seemed like a reasonable compromise that would make integration with engines like Three.js or Babylon clearer without too much overhead.

Given that the matrix representation required some math that we didn't want every developer to have to reproduce (probably incorrectly), and that matricies were more likely to be fed into shader uniforms directly, we did include a special utility attribute for getting a `Float32Array` matrix from an `XRRigidTransform`. It should be noted that the spec allows for the various members of the `XRRigidTransform` to be lazily computed, so an ideal browser implementation won't allocate the `DOMPoint`s or `Float32Array` at all unless they're explicitly asked for.

Closing this issue, since I think the core question of "why" has been answered.
